File: brain-move-web/src/engine/difficulty.ts
export class Difficulty {
  level: number
  history: Array<["up" | "down" | "keep", number]> = []
  private minLevel: number
  private maxLevel: number

  constructor(initialLevel = 1, minLevel = 1, maxLevel = 5) {
    this.level = initialLevel
    this.minLevel = minLevel
    this.maxLevel = maxLevel
  }

  increase(): void {
    if (this.level < this.maxLevel) this.level++
    this.history.push(["up", this.level])
  }

  decrease(): void {
    if (this.level > this.minLevel) this.level--
    this.history.push(["down", this.level])
  }

  keep(): void {
    this.history.push(["keep", this.level])
  }

  adapt(accuracy: number, fatigueLevel: string = "bassa"): void {
    if (accuracy > 0.8 && fatigueLevel === "bassa") this.increase()
    else if (accuracy < 0.5 || fatigueLevel === "alta") this.decrease()
    else this.keep()
  }

  getLevel(): number { return this.level }
}

File: brain-move/engine/difficulty.py
class Difficulty:
    def __init__(self, initial_level=1, min_level=1, max_level=5):
        self.level = initial_level
        self.min_level = min_level
        self.max_level = max_level
        self.history = []

    def increase(self):
        if self.level < self.max_level:
            self.level += 1
        self.history.append(("up", self.level))

    def decrease(self):
        if self.level > self.min_level:
            self.level -= 1
        self.history.append(("down", self.level))

    def keep(self):
        self.history.append(("keep", self.level))

    def adapt(self, accuracy, fatigue_level="bassa"):
        if accuracy > 0.8 and fatigue_level == "bassa":
            self.increase()
        elif accuracy < 0.5 or fatigue_level == "alta":
            self.decrease()
        else:
            self.keep()

    def get_level(self):
        return self.level
